python/cugraph-equivariant/cugraph_equivariant/nn/tensor_product_conv.py

```diff
@@ -0,0 +1,252 @@
+    .. math::
+        \psi_{a b} = \operatorname{MLP}
+        \left(e_{a b}, \mathbf{h}_a^0, \mathbf{h}_b^0\right)
+
+    Parameters
+    ----------
+    in_irreps : e3nn.o3.Irreps
+        Irreps for the input node features.
+
+    sh_irreps : e3nn.o3.Irreps
+        Irreps for the spherical harmonic representations of edge vectors.
+
+    out_irreps : e3nn.o3.Irreps
+        Irreps for the output.
+
+    batch_norm : bool, optional (default=True)
+        If true, batch normalization is applied.
+
+    mlp_channels : sequence of ints, optional (default=None)
+        A sequence of integers defining number of neurons in each layer in MLP
+        before the output layer. If `None`, no MLP will be added.
+
+    mlp_activation : callable, optional (default=torch.nn.GELU)
+        Activation function in MLP.
+
+    mlp_fast_first_layer : bool, optional (default=False)
+        When enabled, for the first layer of MLP, the module performs GEMMs on
+        individual components (i.e., edge embeddings, scalar features of source
+        and destinations nodes) of the weights before indexing and concatenation,
+        leading to a lower complexity in most use cases. This option requires
+        users to explicitly pass in `src_scalars` and `dst_scalars` in
+        `forward()` call.
+
+    use_e3nn_tp: bool, optional (default=False)
+        If `True`, use TensorProduct functions from e3nn.
+
+    Examples
+    --------
+    >>> # Case 1: MLP with the input layer having 6 channels and 2 hidden layers
+    >>> #         having 16 channels. edge_emb.size(1) must match the size of
+    >>> #         the input layer: 6
+    >>>
+    >>> conv1 = FullyConnectedTensorProductConv(in_irreps, sh_irreps, out_irreps,
+    >>>     mlp_channels=[6, 16, 16], mlp_activation=nn.ReLU).cuda()
+    >>> out = conv1(src_features, edge_sh, edge_emb, graph)
+    >>>
+    >>> # Case 2: No MLP, edge_emb will be directly used as the tensor product weights
+    >>>
+    >>> conv2 = FullyConnectedTensorProductConv(in_irreps, sh_irreps, out_irreps,
+    >>>     mlp_channels=None).cuda()
+    >>> out = conv2(src_features, edge_sh, edge_emb, graph)
+    >>>
+    >>> # Case 3: Same as case 1 but with `mlp_fast_first_layer=True`. The scalar
+    >>> #         features from edges, sources and destinations have to be passed
+    >>> #         in separately.
+    >>>
+    >>> conv3 = FullyConnectedTensorProductConv(in_irreps, sh_irreps, out_irreps,
+    >>>     mlp_channels=[6, 16, 16], mlp_fast_first_layer=True).cuda()
+    >>> out = conv3(src_features, edge_sh, edge_scalars, graph,
+    >>>     src_scalars=src_scalars, dst_scalars=dst_scalars)
+    """
+
+    def __init__(
+        self,
+        in_irreps: o3.Irreps,
+        sh_irreps: o3.Irreps,
+        out_irreps: o3.Irreps,
+        batch_norm: bool = True,
+        mlp_channels: Optional[Sequence[int]] = None,
+        mlp_activation: Optional[Callable[..., nn.Module]] = nn.GELU,
+        mlp_fast_first_layer: bool = False,
```

Comment:
Why does this argument need to be here? Can't the special just happen when `src_scalars` and `dst_scalars` are provided?

Reply:
I agree with Mario here. There are a few branches when it comes to computing the tp weights:
1. user direct input the precomputed weights to forward
2. feed only the edge_emb to the mlp to compute the weight. This could include the case that the node embeddings are already concatenated into the edge_emb, which we can't distinguish
3. As in point 2 but additionally have one node embedding tensor, i.e., the graph is non-directed
4. As in point 2 but additionally have separate src and dst node embedding tensors

Reply:
I've updated the code to support arbitrary number of scalar features in [src, dst]_scalars. They can also be `None` if needed.

Regarding use case 3, users should input `src_scalars=node_attr, dst_scalars=node_attr`.